Add operations to interact with Ports

CamelCaseNotation opened the pull request:
Add basic Assign/Unassign functionality of virtual networks to
bonded ports

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/packethost/packngo/59)
<!-- Reviewable:end -->

t0mk commented:
Thanks for the PR I will do review on Monday.

t0mk commented:
hi, this PR looks good, but still 
- test is missing
- It would really make sense to also implement the Virtual Networks API (https://www.packet.net/developers/api/virtualnetworks/) together with the ports. How would you feel about doing the VLANs too in this PR? 

If we'd have VLANs, it would be easier to test.

I can help with the VLANs, or with writing test somehow, if you don't feel like doing it. Just let me know, and make sure this PR has "allowed updates from maintainers".

CamelCaseNotation commented:
For the tests I was unsure what the best way would be to go about testing this. I can go ahead and implement the Virtual Networks API as well. Once I update the PR with VLANs could you help me write the tests?

CamelCaseNotation commented:
What do you think about keeping the virtual network implementation in the `ports.go` file? The API endpoint is still `/ports`

CamelCaseNotation commented:
I haven't manually tested these changes like I did for port assign/unassign operations, however once the test cases are written that should provide all the testing necessary.

t0mk commented:
Thanks @CamelCaseNotation. I will write tests for the VLANs and then for the ports this afternoon.

t0mk commented:
@CamelCaseNotation, I noticed that you pack params for simple API request to structs:
https://github.com/CamelCaseNotation/packngo/blob/0899b7eda431bef58e699145241bba8d327402cb/virtualnetworks.go#L11

So far in packngo, we've been just listing the params in function declaration straight (when it was just one or two), like in:
https://github.com/packethost/packngo/blob/master/volumes.go#L11

However, your way might be better, I just asked @mmlb for opinion on this, and I'd like to resolve this before I write the tests.

Thanks a lot for you contritbution!

mmlb commented:
Thanks a lot for the contribution! I have a few nits that need taking care of. Regarding what @t0mk says, I'd like to have the methods that do similar things named similarly and that also requires that they take similar inputs (patterns at least). So the Virtual* funcs should be:
```
List(id)
Delete(id)
Create(VirtualNetworkCreateRequest)
```

---

Reviewed 3 of 5 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 21 unresolved discussions.

---

*[devices.go, line 14 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QDeZxz0BxAoJtr-SK:-L2QDeZyCX_FPx6wH-V7:b8jf91v) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/devices.go#L14)):*
> ```Go
> 	List(ProjectID string) ([]Device, *Response, error)
> 	Get(string) (*Device, *Response, error)
> 	GetWith(string, []string) (*Device, *Response, error)
> ```

Hmm this should be named something else maybe, and the declaration should be `$Name(string, include, exclude []string)`

---

*[devices.go, line 50 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QFRAiUb2BuQ1uwwwa:-L2QFRAiUb2BuQ1uwwwb:b-uw87y0) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/devices.go#L50)):*
> ```Go
> 	ProvisionPer        float32                `json:"provisioning_percentage,omitempty"`
> 	UserData            string                 `json:"userdata,omitempty"`
> 	NetworkPorts        []Port                 `json:"network_ports"`
> ```

should be at the bottom of the list

---

*[devices.go, line 138 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QFaoI9oA85W4HT5-D:-L2QFaoJo4swbrp74b-X:b6ol27v) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/devices.go#L138)):*
> ```Go
> 
> // Get returns a device by id
> func (s *DeviceServiceOp) Get(deviceID string) (*Device, *Response, error) {
> ```

it would be good to refactor Get to call GetWith

---

*[ports.go, line 29 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QG6JPi7176Dh_HncQ:-L2QG6JPi7176Dh_HncR:bnq7v3s) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/ports.go#L29)):*
> ```Go
> }
> 
> type PortAssignInput struct {
> ```

Lets follow the rest of the package and name this `PortAssignRequest`, here and everywhere else in the file

---

*[ports.go, line 33 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QG1k_5EZPCYpBTyVa:-L2QG1k_5EZPCYpBTyVb:b4j1oxx) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/ports.go#L33)):*
> ```Go
> 	PortId           string
> 	VirtualNetworkId int `json:"vnid"`
> }
> ```

Id -> ID, throughout this file

---

*[ports.go, line 35 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QGGpN6qlNZajnyFEQ:-L2QGGpN6qlNZajnyFER:b6mxjuj) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/ports.go#L35)):*
> ```Go
> }
> 
> type PortAssignOutput struct {
> ```

rename to PortAssignResponse here and everywhere else

---

*[ports.go, line 47 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QGQ8HJLLSfKt6GIrf:-L2QGQ8HJLLSfKt6GIrg:by6mv64) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/ports.go#L47)):*
> ```Go
> 
> 	// No network ports for this device so no-op
> 	if len(device.NetworkPorts) == 0 {
> ```

you can get rid of this check, the return after the loop has the same affect.

---

*[ports.go, line 50 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QGlgHeLDDxr-sX5r6:-L2QGi7l83xp84MWxrsi:bwiy1gy) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/ports.go#L50)):*
> ```Go
> 		return nil, nil, false, err
> 	}
> 	for index := range device.NetworkPorts {
> ```

can you rewrite as
```
for _, port := ...
if port.ID ... {
continue
}

if len(port...) == 0 {
...
else 
...
}
}
```

---

*[ports.go, line 55 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QHCMVlFQIgCwuHMib:-L2QHCMVlFQIgCwuHMic:b-yiqbnv) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/ports.go#L55)):*
> ```Go
> 		} else {
> 			if len(port.AttachedVirtualNetworks) == 0 {
> 				// convert to layer-3 (and attach vlan)
> ```

this comment contradicts the comment at the beginning of the function and what `i.convertLayerTwo` is seemingly doing

---

*[ports.go, line 98 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QLDpwGdyOmQr_qHPS:-L2QLDpwGdyOmQr_qHPT:b5is4q7) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/ports.go#L98)):*
> ```Go
> }
> 
> func (i *DevicePortServiceOp) Bond(input *PortBondInput) (*PortBondOutput, *Response, error) {
> ```

I think returning just `Port` should be fine

---

*[ports.go, line 104 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QHvDS_EY8UqHpXY_n:-L2QHvDS_EY8UqHpXY_o:br1rkzi) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/ports.go#L104)):*
> ```Go
> 	} else {
> 		path = fmt.Sprintf("%s/%s/bond", portBasePath, input.PortId)
> 	}
> ```

I'd prefer:
```
path = fmt.Sprintf()
if input.BulkEnable {
    path += ...
}
```

---

*[ports.go, line 124 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QLIVQQUIj7kLqhZEG:-L2QLIVQQUIj7kLqhZEH:byfwm71) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/ports.go#L124)):*
> ```Go
> }
> 
> func (i *DevicePortServiceOp) Disbond(input *PortDisbondInput) (*PortDisbondOutput, *Response, error) {
> ```

ditto re port return

---

*[ports.go, line 124 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QLQeU6RYnevUgfIqS:-L2QLQeU6RYnevUgfIqT:b79oyzs) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/ports.go#L124)):*
> ```Go
> }
> 
> func (i *DevicePortServiceOp) Disbond(input *PortDisbondInput) (*PortDisbondOutput, *Response, error) {
> ```

This can just be input params of `portID string, bulkDisable bool`

---

*[ports.go, line 143 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QII19qxHuFoLZCevl:-L2QII1Apnv7jSccNQIn:b-lpwz94) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/ports.go#L143)):*
> ```Go
> func (i *DevicePortServiceOp) GetBondedPort(deviceId string) (*Port, bool, error) {
> 	device, _, err := i.client.Devices.Get(deviceId)
> 	if len(device.NetworkPorts) == 0 {
> ```

can drop this branch

---

*[ports.go, line 147 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QIJsuwGf66q01FniB:-L2QIJsuwGf66q01FniC:b-l9n9kf) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/ports.go#L147)):*
> ```Go
> 	}
> 	for index := range device.NetworkPorts {
> ```

extra space

---

*[ports.go, line 168 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QIRkayUeeSLxJ6zhn:-L2QIRkayUeeSLxJ6zho:b-be45ld) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/ports.go#L168)):*
> ```Go
> }
> 
> func (i *DevicePortServiceOp) convertLayerTwo(input *PortAssignInput) (*PortAssignOutput, *Response, bool, error) {
> ```

I think covertToLayer2 is a better name, it better describes the action performed

---

*[virtualnetworks.go, line 32 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QImMnBXFKYYJd_Gn4:-L2QImMnBXFKYYJd_Gn5:bpcs8es) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/virtualnetworks.go#L32)):*
> ```Go
> type VirtualNetworkListInput struct {
> 	ProjectId string
> 	Includes  []string
> ```

Is anything using the Includes? I'd rather keep all the `List` and `Delete` looking the same for all the different services. I know we need a better method of doing `include` and `exclude`, but that should be done in a method named differently.

---

*[virtualnetworks.go, line 59 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QJBMqnAIzpHgjDAwh:-L2QJBMqnAIzpHgjDAwi:b32vvk) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/virtualnetworks.go#L59)):*
> ```Go
> 
> type VirtualNetworkCreateInput struct {
> 	ProjectId   string `json:"project_id"`
> ```

ID

---

*[virtualnetworks.go, line 62 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QJC_6odfYkXj_rkRL:-L2QJC_7K-lRe3AUquPZ:b-7zpjla) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/virtualnetworks.go#L62)):*
> ```Go
> 	Description string `json:"description"`
> 	Facility    string `json:"facility"`
> 	Vxlan       int    `json:"vxlan"`
> ```

VXLAN

---

*[virtualnetworks.go, line 63 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QJE9L1YRfwth0u3WD:-L2QJE9L1YRfwth0u3WE:bb75drc) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/virtualnetworks.go#L63)):*
> ```Go
> 	Facility    string `json:"facility"`
> 	Vxlan       int    `json:"vxlan"`
> 	Vlan        int    `json:"vlan"`
> ```

VLAN

---

*[virtualnetworks.go, line 71 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QJHDtg1sosa3FWGRd:-L2QJHDtg1sosa3FWGRe:bug3gm5) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/virtualnetworks.go#L71)):*
> ```Go
> 
> func (i *ProjectVirtualNetworkServiceOp) Create(input *VirtualNetworkCreateInput) (*VirtualNetworkCreateOutput, *Response, error) {
> 	// TODO: May need to add timestamp to output from 'post' request
> ```

I'll check if this is an api bug

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/packethost/packngo/59#-:-L2QJNpdn2_qwb3KGbh7:b-qqxnn3)*
<!-- Sent from Reviewable.io -->

CamelCaseNotation commented:
Thanks for reviewing! I'll address the comments directly via Reviewable. Regarding the method input types, I picked up this pattern from using the AWS SDK. What it does is prevents future refactoring since you can add additional "arguments" by extending the struct fields.

I realize this could be overkill given the API might not change much. It would also solve how to handle any query parameter stuff like "bulk_enable" or "include" by simply setting the field in the input struct to be nil when creating the request.

Using the structs for every method gives consistency as well, although I do admit it's unnecessarily verbose when you just need one argument. 

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/packethost/packngo/59#-:-L2Sfqav7LyC12_1IVkt:b-ku3gse)*
<!-- Sent from Reviewable.io -->

CamelCaseNotation commented:
*[devices.go, line 14 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QDeZxz0BxAoJtr-SK:-L2SiWiZNwBTshFWEVU4:b-5vhsyc) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/devices.go#L14)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

Hmm this should be named something else maybe, and the declaration should be `$Name(string, include, exclude []string)`
</blockquote></details>

I was trying to avoid a long name like "GetAndInclude", but "GetWith" isn't that clear either. Any suggestions? What is the exclude string slice used for?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/packethost/packngo/59)*
<!-- Sent from Reviewable.io -->

CamelCaseNotation commented:
Review status: all files reviewed at latest revision, 21 unresolved discussions.

---

*[ports.go, line 50 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QGlgHeLDDxr-sX5r6:-L2SkvSCvSGxTsTBprn1:b-j5484i) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/ports.go#L50)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

can you rewrite as
```
for _, port := ...
if port.ID ... {
continue
}

if len(port...) == 0 {
...
else 
...
}
}
```
</blockquote></details>

Oops, I forgot my idiomatic Go patterns!! Thank you, this is much cleaner

---

*[ports.go, line 55 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QHCMVlFQIgCwuHMib:-L2Sm9M-0WDkMKUUIzgt:b-dll6hj) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/ports.go#L55)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

this comment contradicts the comment at the beginning of the function and what `i.convertLayerTwo` is seemingly doing
</blockquote></details>

This was a typo, fixed

---

*[ports.go, line 98 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QLDpwGdyOmQr_qHPS:-L2SmFGvv7LId_OYolCJ:b-1r7vfs) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/ports.go#L98)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

I think returning just `Port` should be fine
</blockquote></details>

Agreed, but see top-level comment about why I chose to use structs

---

*[ports.go, line 104 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QHvDS_EY8UqHpXY_n:-L2Sm_wWU1iC4G97Ed4d:b-mmceih) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/ports.go#L104)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

I'd prefer:
```
path = fmt.Sprintf()
if input.BulkEnable {
    path += ...
}
```
</blockquote></details>

Oh yeah, this works nicely since the query string will always be an addition to the end :thumbsup: 

---

*[ports.go, line 168 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QIRkayUeeSLxJ6zhn:-L2SnWR_-Eu4D65Z7mbr:b3p9cyw) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/ports.go#L168)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

I think covertToLayer2 is a better name, it better describes the action performed
</blockquote></details>

Changed to `convertToLayerTwo` since I think numbers in function/method names are icky at times

---

*[virtualnetworks.go, line 32 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QImMnBXFKYYJd_Gn4:-L2Snm-_mbBN6tjYF7x2:bhwg4vd) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/virtualnetworks.go#L32)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

Is anything using the Includes? I'd rather keep all the `List` and `Delete` looking the same for all the different services. I know we need a better method of doing `include` and `exclude`, but that should be done in a method named differently.
</blockquote></details>

I plan to use these changes to extend the Packet provider plugin for Terraform, so there may be a need to use the Includes. With the struct approach, should I need it, it's just a matter of adding what I want as the `Includes` field vs having to refactor the API/add a new method in order to provide that functionality.

---

*[virtualnetworks.go, line 59 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QJBMqnAIzpHgjDAwh:-L2So8qYLbBiROBg8b-h:b53uwdv) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/virtualnetworks.go#L59)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

ID
</blockquote></details>

ID is misleading since it requires a project ID and having to reference the struct tag to know that vs looking at the field name seems bad.

---

*[virtualnetworks.go, line 71 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QJHDtg1sosa3FWGRd:-L2SofWSoZpNnm2lQCH9:b-4apx2d) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/virtualnetworks.go#L71)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

I'll check if this is an api bug
</blockquote></details>

This comment was based on what the output looks like on the Packet API documentation page: https://www.packet.net/developers/api/virtualnetworks/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/packethost/packngo/59)*
<!-- Sent from Reviewable.io -->

mmlb commented:
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[devices.go, line 14 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QDeZxz0BxAoJtr-SK:-L2VZU4hg-NlQ1TOTUob:b-yyzhgw) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/devices.go#L14)):*
<details><summary><i>Previously, CamelCaseNotation (Joshua Hurt) wrote…</i></summary><blockquote>

I was trying to avoid a long name like "GetAndInclude", but "GetWith" isn't that clear either. Any suggestions? What is the exclude string slice used for?
</blockquote></details>

:thinking: 

GetExtra
GetCustom
GetCluded

I'll take an internal poll see if anyone has a good idea.

I'm not sure if there's a use for it from the public api, but internally we have some responses that return way more data than is normally cared about, so we have $foo_lite include option and we combine it with exclude=$foo. There's no reason why this won't be of use publicly.

FYI: I heavily feel the issue of lack of good include/exclude handling. Most of our internal go services have their own custom packet client (each!) that basically just builds requests and calls http methods, its a PITA as you can imagine. They each care about specific include/excludes so thats the reason for not sharing. I had something started that allowed include/exclude + private apis, I'll look around for the include/exclude stuff.

---

*[ports.go, line 98 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QLDpwGdyOmQr_qHPS:-L2VY5DgWNVWMXzTcaRd:bc9xl29) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/ports.go#L98)):*
<details><summary><i>Previously, CamelCaseNotation (Joshua Hurt) wrote…</i></summary><blockquote>

Agreed, but see top-level comment about why I chose to use structs
</blockquote></details>

I agree with you, but we already have an established api pattern and I'd rather have a ugly but consistent api vs ugly and inconsistent :P. 

---

*[ports.go, line 168 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QIRkayUeeSLxJ6zhn:-L2VXFj7CUxYB2Iy7ilq:b-b7mykj) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/ports.go#L168)):*
<details><summary><i>Previously, CamelCaseNotation (Joshua Hurt) wrote…</i></summary><blockquote>

Changed to `convertToLayerTwo` since I think numbers in function/method names are icky at times
</blockquote></details>

I don't mind them, but `Two` isn't a problem either.

---

*[virtualnetworks.go, line 32 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QImMnBXFKYYJd_Gn4:-L2VYHh8m1fCTo0WilMM:b-67tqgh) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/virtualnetworks.go#L32)):*
<details><summary><i>Previously, CamelCaseNotation (Joshua Hurt) wrote…</i></summary><blockquote>

I plan to use these changes to extend the Packet provider plugin for Terraform, so there may be a need to use the Includes. With the struct approach, should I need it, it's just a matter of adding what I want as the `Includes` field vs having to refactor the API/add a new method in order to provide that functionality.
</blockquote></details>

Same deal with consistency argument. I'd like to have a way to specify Include/Exclude too, but I don't want to have 3 similar methods, normal List, this structed List, and ListExtra that gets added for the previous services.

FYI we (well @tomk :p) are working on a new client thats autogenerated from swagger spec. Not sure how far along that is or how it'll look like in the end. Either way there's probably a future where this version is archived in favor of swagger or packngo v2 that will try to fix the very roughness of this api.

---

*[virtualnetworks.go, line 59 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QJBMqnAIzpHgjDAwh:-L2VXZoy8L-96T5ke9Oi:b-bl54gr) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/virtualnetworks.go#L59)):*
<details><summary><i>Previously, CamelCaseNotation (Joshua Hurt) wrote…</i></summary><blockquote>

ID is misleading since it requires a project ID and having to reference the struct tag to know that vs looking at the field name seems bad.
</blockquote></details>

You are correct, I was unclear in that I wanted you to do the Id -> ID xform, :+1:

---

*[virtualnetworks.go, line 71 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QJHDtg1sosa3FWGRd:-L2VaDKheRPzNMu8nvHP:b-25n17x) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/virtualnetworks.go#L71)):*
<details><summary><i>Previously, CamelCaseNotation (Joshua Hurt) wrote…</i></summary><blockquote>

This comment was based on what the output looks like on the Packet API documentation page: https://www.packet.net/developers/api/virtualnetworks/
</blockquote></details>

oic, why is the created_at here then?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/packethost/packngo/59)*
<!-- Sent from Reviewable.io -->

TomK commented:
@mmlb not me, you mean @t0mk 
odd though, i was discussing go client generation from swagger with a colleague just this morning o.0

mmlb commented:
@TomK spooky

CamelCaseNotation commented:
@TomK :ghost: :ghost: 

---

Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[devices.go, line 14 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QDeZxz0BxAoJtr-SK:-L2WmlujpWG_RRxTDcSv:bkfuhwj) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/devices.go#L14)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

:thinking: 

GetExtra
GetCustom
GetCluded

I'll take an internal poll see if anyone has a good idea.

I'm not sure if there's a use for it from the public api, but internally we have some responses that return way more data than is normally cared about, so we have $foo_lite include option and we combine it with exclude=$foo. There's no reason why this won't be of use publicly.

FYI: I heavily feel the issue of lack of good include/exclude handling. Most of our internal go services have their own custom packet client (each!) that basically just builds requests and calls http methods, its a PITA as you can imagine. They each care about specific include/excludes so thats the reason for not sharing. I had something started that allowed include/exclude + private apis, I'll look around for the include/exclude stuff.
</blockquote></details>

Interesting, could you provide a more concrete example of the include/exclude use case? Or perhaps an example of using exclude in a public API call? For this function signature I think it would make sense to have a `GetCustomRequest` struct, maybe?? In this case perhaps not, unless `Get` were to be refactored cuz otherwise you'd have to build a transitive struct to allow `Get` to call `GetExtra` (or whatever the poll results are :P)

---

*[ports.go, line 47 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QGQ8HJLLSfKt6GIrf:-L2WotD_MEsGytgz35kw:b-45ls8c) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/ports.go#L47)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

you can get rid of this check, the return after the loop has the same affect.
</blockquote></details>

I don't understand, what do you mean? I can completely remove the if statement and leave it as is? 

---

*[ports.go, line 98 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QLDpwGdyOmQr_qHPS:-L2X3lb7ujoqFk-FWZl6:b-tbmeec) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/ports.go#L98)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

I agree with you, but we already have an established api pattern and I'd rather have a ugly but consistent api vs ugly and inconsistent :P. 
</blockquote></details>

Very good point!! Consistently ugly :P 

---

*[virtualnetworks.go, line 32 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QImMnBXFKYYJd_Gn4:-L2X4iP-yOZLLUHTHs4e:b-gmychf) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/virtualnetworks.go#L32)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

Same deal with consistency argument. I'd like to have a way to specify Include/Exclude too, but I don't want to have 3 similar methods, normal List, this structed List, and ListExtra that gets added for the previous services.

FYI we (well @tomk :p) are working on a new client thats autogenerated from swagger spec. Not sure how far along that is or how it'll look like in the end. Either way there's probably a future where this version is archived in favor of swagger or packngo v2 that will try to fix the very roughness of this api.
</blockquote></details>

So what you're saying is this whole API will be stubbed out from the swagger-go generator after consuming a swagger spec? Make it a swagger 3.0 spec if you still can!! 

---

*[virtualnetworks.go, line 59 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QJBMqnAIzpHgjDAwh:-L2X4x_i36Shv_oYKMat:b8l87b5) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/virtualnetworks.go#L59)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

You are correct, I was unclear in that I wanted you to do the Id -> ID xform, :+1:
</blockquote></details>

OH, I misread that a bit haha

---

*[virtualnetworks.go, line 71 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QJHDtg1sosa3FWGRd:-L2X56ZYwHMmgR6FnuHy:btauld) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/virtualnetworks.go#L71)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

oic, why is the created_at here then?
</blockquote></details>

Cuz when you look at the output of virtual networks in, for example, making a GET request against the devices endpoint while specifying `?include=virtual_networks` the JSON has the `created_at` attribute. This is why I thought it was curious that in the API docs under `201 response` for the POST the JSON doesn't include `created_at`. I haven't actually performed that API call yet to confirm `created_at` is omitted.

I actually had a test file I was using to locally test my API changes, but it involved moving all the top-level `.go` files into a directory named `packngo` and then importing that via relative pathing so that my local API changes were picked up. What a big pain in the butt!!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/packethost/packngo/59#-:-L2WVSCV5LGpZlmtwPQm:b-j3o0f2)*
<!-- Sent from Reviewable.io -->

CamelCaseNotation commented:
Review status: 5 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

*[devices.go, line 14 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QDeZxz0BxAoJtr-SK:-L2XOeMaoQ4HlgTkvDB8:b-m8rx19) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/devices.go#L14)):*
<details><summary><i>Previously, CamelCaseNotation (Joshua Hurt) wrote…</i></summary><blockquote>

Interesting, could you provide a more concrete example of the include/exclude use case? Or perhaps an example of using exclude in a public API call? For this function signature I think it would make sense to have a `GetCustomRequest` struct, maybe?? In this case perhaps not, unless `Get` were to be refactored cuz otherwise you'd have to build a transitive struct to allow `Get` to call `GetExtra` (or whatever the poll results are :P)
</blockquote></details>

Is include and exclude mutually exclusive for one particular request? Or can you have both `?include=<csv's>?exclude=<csv's>`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/packethost/packngo/59)*
<!-- Sent from Reviewable.io -->

mmlb commented:
Review status: 4 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

*[ports.go, line 47 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QGQ8HJLLSfKt6GIrf:-L2aELesPHt-yxq7cSDk:bt8bwfc) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/ports.go#L47)):*
<details><summary><i>Previously, CamelCaseNotation (Joshua Hurt) wrote…</i></summary><blockquote>

I don't understand, what do you mean? I can completely remove the if statement and leave it as is? 
</blockquote></details>

I thought I marked this as "satisfied" already...? You addressed it already when you removed the `if len(device.NetworkPorts) == 0 {...}` block

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/packethost/packngo/59)*
<!-- Sent from Reviewable.io -->

CamelCaseNotation commented:
Review status: 4 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[ports.go, line 47 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QGQ8HJLLSfKt6GIrf:-L2fqvhkBQ4to0bf1CZW:b-wnry65) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/ports.go#L47)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

I thought I marked this as "satisfied" already...? You addressed it already when you removed the `if len(device.NetworkPorts) == 0 {...}` block
</blockquote></details>

I redid it a little anyways to look nicer

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/packethost/packngo/59)*
<!-- Sent from Reviewable.io -->

CamelCaseNotation commented:
@t0mk Should the test cases be written with the assumption a valid API key and consumer token will be provided? I assume Packet has a key/token and facility to be used for this repo to use for test cases.

mmlb commented:
Its been on my TODO list to keep bugging someone to get a project setup that we can use in CI. For now I/t0mk  run manually against our keys. A mock/test facility is not needed for these tests.

---

Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[devices.go, line 14 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QDeZxz0BxAoJtr-SK:-L2aEzTAo6ofPy5pXID6:brka3zu) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/devices.go#L14)):*
<details><summary><i>Previously, CamelCaseNotation (Joshua Hurt) wrote…</i></summary><blockquote>

Is include and exclude mutually exclusive for one particular request? Or can you have both `?include=<csv's>?exclude=<csv's>`?
</blockquote></details>

<shame>I forgot to take the poll</shame>
Yeah I'm pretty sure I can whip up an example later today.
I think its ok to have 3 params as is, though its right on the fence I think.
You can have both, I make use of `include=project_lite,exclude=project` in one of our services

...

I forgot to Publish :(
Meager poll results came back, only GetCluded showed any interest but I don't really like it though it kind of makes the most sense. I'll let you decide.

---

*[devices.go, line 148 at r4](https://reviewable.io:443/reviews/packethost/packngo/59#-L2zek4Kz2_U7vZoEIs9:-L2zek4Kz2_U7vZoEIsA:b58j9es) ([raw file](https://github.com/packethost/packngo/blob/d3e1cd67430102fd21408fe66a74108acdc478ca/devices.go#L148)):*
> ```Go
> 	if includes != nil {
> 		path += fmt.Sprintf("?include=%s", strings.Join(includes, ","))
> 	} else if excludes != nil {
> ```

these should not be mutually exclusive

---

*[virtualnetworks.go, line 32 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QImMnBXFKYYJd_Gn4:-L2ze0BOUqPqptl_jfXw:b-ngp3yu) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/virtualnetworks.go#L32)):*
<details><summary><i>Previously, CamelCaseNotation (Joshua Hurt) wrote…</i></summary><blockquote>

So what you're saying is this whole API will be stubbed out from the swagger-go generator after consuming a swagger spec? Make it a swagger 3.0 spec if you still can!! 
</blockquote></details>

@t0mk is doing all the work on that front.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/packethost/packngo/59#-:-L2zeGShODwON6gHDw7B:bejq4cu)*
<!-- Sent from Reviewable.io -->

CamelCaseNotation commented:
Is it okay to use the consumer token from the terraform packet provider for testing purposes?`aZ9GmqHTPtxevvFq9SK3Pi2yr9YCbRzduCSXF2SNem5sjB91mDq7Th3ZwTtRqMWZ`

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[devices.go, line 14 at r2](https://reviewable.io:443/reviews/packethost/packngo/59#-L2QDeZxz0BxAoJtr-SK:-L30xr0A7tZP75Sif0Oy:b9k30tn) ([raw file](https://github.com/packethost/packngo/blob/263c53ac8214500dcc3ff1de558ca0fedd99712a/devices.go#L14)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

<shame>I forgot to take the poll</shame>
Yeah I'm pretty sure I can whip up an example later today.
I think its ok to have 3 params as is, though its right on the fence I think.
You can have both, I make use of `include=project_lite,exclude=project` in one of our services

...

I forgot to Publish :(
Meager poll results came back, only GetCluded showed any interest but I don't really like it though it kind of makes the most sense. I'll let you decide.
</blockquote></details>

I'll go for verbosity than cleverness, how about `GetIncludeExclude`

---

*[devices.go, line 148 at r4](https://reviewable.io:443/reviews/packethost/packngo/59#-L2zek4Kz2_U7vZoEIs9:-L30xbBNM_0_W7iw-B5O:bth82d2) ([raw file](https://github.com/packethost/packngo/blob/d3e1cd67430102fd21408fe66a74108acdc478ca/devices.go#L148)):*
<details><summary><i>Previously, mmlb (Manuel Mendez) wrote…</i></summary><blockquote>

these should not be mutually exclusive
</blockquote></details>

Okay I wasn't sure of the usage, lemme update it

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/packethost/packngo/59#-:-L30z5jCbmXe4SH-FVEx:blrs6d)*
<!-- Sent from Reviewable.io -->

CamelCaseNotation commented:
I'm wanting to go ahead and write `test_ports` and `test_virtualnetworks` but my company has no more devices available for me to create in our projects, would you be able to provide one? Perhaps a sandbox one? Or maybe you could answer some questions for me as I'm not terribly familiar with all the packet flow.

When I create a device, are there any ports for it? If so, will a bonded port already be created when I create a device?

I'm wanting to do the following:
- create project with `setupWithProject()`
- create device (assuming at least one port will be made and attached to this device)
- assert project has one device
- enable bonding for a port on that device
- use deviceID to get the bonded port
- assert bonded port exists and has no IPs or virtual networks attached to it
- create a virtual network for the project
- use the vxlan as input to a port `Assign`
- assert port has 1 virtual network
- `Unassign` the virtual network
- assert port has zero virtual networks
- delete virtual network
- assert project has zero virtual networks
- disbond the port
- assert device has zero bonded ports
- cleanup

Thoughts? I think this should test all the functions I added

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/packethost/packngo/59#-:-L311APkKwHWizYfVqTU:b-nubs8x)*
<!-- Sent from Reviewable.io -->

CamelCaseNotation commented:
This is the "everything works" flow. There would also be checks sprinkled in to test erroneous behavior doesn't go through such as attempting to `Assign` a virtual network to a port where it's already assigned

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/packethost/packngo/59#-:-L315PudnwoUrXakk_Tv:b-k5g1jz)*
<!-- Sent from Reviewable.io -->

t0mk commented:
@CamelCaseNotation I added a test for virtualnetworks and fixed it a bit, check out https://github.com/packethost/packngo/pull/59/commits/8a64a2712d616a68c847b823ef8217370595571e. Let me know if you identify something horrible in it. Feel free to fix it yourself.

I will go on and write test for the ports.go.

As you probably noticed, for packngo we have only acceptance test - tests that run against real API. The test usually creates disposable project:
```
c, projectID, teardown := setupWithProject(t)
defer teardown()
```
.. and then creates tested resources in the project. After the test, all the resources are removed. Most of the time, the test really cleans after itself. Resources are allocated for a couple of minutes - you get charged very little.

t0mk commented:
@CamelCaseNotation if you'd like to run the VLAN test yourself, you could do
```
$ PACKNGO_TEST_FACILITY=sin1 PACKNGO_TEST_ACTUAL_API=1 go test -v -run=TestAccVirtualNetworks
```

If you want to see the HTTP messages, you can do
```
PACKNGO_DEBUG=1 PACKNGO_TEST_FACILITY=sin1 PACKNGO_TEST_ACTUAL_API=1 go test -v -run=TestAccVirtualNetworks
```

t0mk commented:
ah, I see that L2 is only compatible for "Type 2, Type 2A, and Type 3 servers, and our new Type S storage server.". That will be fun to test..

CamelCaseNotation commented on virtualnetworks_test.go:
```diff
@@ -1,21 +1,69 @@
+	}
+
+	if len(l.VirtualNetworks) != 1 {
+		t.Fatal("Newly created project should not have any vlans")
```
This message should say "Newly created project should have exactly one vlan"

CamelCaseNotation commented:
@t0mk What happens if you attempt to attach a virtual network to a bonded port on a server type that isn't one of the ones you listed? Is if even possible to have a bonded port on such a server type?

We _could_ check `plan.name` to see if it matches one of those four you listed, but I don't like that since the list of types of servers would need to be updated as y'all add L2 support for more/add more server types

t0mk commented:
@CamelCaseNotation I don't exactly know what happens when you try to attach VLAN to L2-unsupperted sever type. I am hoping for some sensible HTTP 4/5xx error. I am against checking the server type on assign - if it blows, it should be fixed on the API side IMO. 

I will try that next week and let you know. I will also write the ports test next week, maybe we will learn more from it. I am now home with flu.

CamelCaseNotation commented:
I hope you feel better soon @t0mk! Get plenty of rest. If I get some time this weekend I'll write the port test cases and push them here

t0mk commented:
@CamelCaseNotation thanks, I feel better now.

I rewrote the ports.go quite drastically. I tried to make it more simple. I also wrote a test for Type 1E.

Now I try to write a test for Type 2, as soon as I find a facility with free slot where L2 is supported.

t0mk commented:
@CamelCaseNotation, I wrote test for other machine types. Could you please go through the test function from here (line 174) till the end:
https://github.com/CamelCaseNotation/packngo/blob/0de54627607493bbb3c852c67f0f48355b6b1f92/ports_test.go#L174
.. and check if the test make any sense?

CamelCaseNotation commented on ports.go:
```diff
@@ -26,140 +28,105 @@ type DevicePortServiceOp struct {
-type PortDisbondRequest struct {
-	PortID      string
-	BulkDisable bool
+func (i *DevicePortServiceOp) assignmentAction(portID, vlanID, path string) (*Port, *Response, error) {
```
I'm not a fan of using this for both `Assign` and `Unassign` in the event that should you ever need to change the logic of how one of these two functions work then it makes it difficult since they follow the same logic path

t0mk replied:
I got rid of the assignmentAction and kept only portAction. It really saves space to do the API request for these 4 similar tasks in one method, and it's kept readable I think.

CamelCaseNotation commented on ports.go:
```diff
@@ -26,140 +28,105 @@ type DevicePortServiceOp struct {
-func (i *DevicePortServiceOp) assignVirtualNetwork(input *PortAssignRequest) (*PortAssignResponse, *Response, bool, error) {
-	path := fmt.Sprintf("%s/%s/assign", portBasePath, input.PortID)
-	assignResponse := new(PortAssignResponse)
+func (i *DevicePortServiceOp) portAction(path string, req interface{}) (*Port, *Response, error) {
```
By moving the struct abstraction to the internals of the API call, you're removing the reason to use a struct in the first place: flexibility and ease of refactoring required arguments later. If you need to add an additional argument to any of these API calls, you will not only have to refactor all the code that calls these functions, but also have to refactor the struct to add an additional field, and unpack the arguments into the struct within the function.

t0mk replied:
I put it back to structs in 3c2b87a9dfde24d7d7f1126ee55aaa5ae8fee390

CamelCaseNotation commented on ports.go:
```diff
@@ -26,140 +28,105 @@ type DevicePortServiceOp struct {
-func (i *DevicePortServiceOp) convertToLayerTwo(input *PortAssignRequest) (*PortAssignResponse, *Response, bool, error) {
-	path := fmt.Sprintf("%s/%s/convert/layer-2", portBasePath, input.PortID)
-	assignResponse := new(PortAssignResponse)
+func (i *DevicePortServiceOp) ConvertToLayerTwo(portID string) (*Port, *Response, error) {
```
Breaking this out into another distinct function makes sense, but it then also means every application has to implement their own logic for deciding when a conversion to L2 is necessary before calling `Assign`. If this is going to stay a distinct function call, then there should be another function that is called `AssignAndConvert`

t0mk replied:
The type 1E has L2 support on "eth1" interface, but doesn't need to be converted to Layer 2, I split it because of this.

t0mk commented:
I got some more info from Packet people, about how the modes (L3, L2, L2hybrid) are treated from API point of view. 

I will write wrapper methods for conversions of devices to L2, L2hybrid and L3 (a bit like @CamelCaseNotation had in the beginning, but for all of the modes), and then I write tests for it.

The testing takes long time, and this PR too, but let's try to finish it!

t0mk commented:
I finished the API for switching in network modes, see top of ports.go for details.

I also wrote tests for most of the new code. I am merging this PR. Let's tackle details in other issues, this PR is too big already anyway.

CamelCaseNotation commented:
@t0mk The portAction function is clean! Before I think there was some packing of variables into the structs and it felt clunky. Great work on figuring out the network type situation, thank you!

I'm glad this got merged, it's been a while but it was worth!
